proposals/NNNN-explicit-padding.md

```diff
@@ -0,0 +1,175 @@
+---
+title: "[NNNN] - Explicit Padding in Structs and CBuffer Arrays"
+params:
+    status: Design In Progress
+    authors:
+        - bogner: Justin Bogner
+---
+
+
+## Introduction
+
+We introduce an explicit padding type for HLSL, and construct cbuffer arrays
+and structs that are annotated with `packoffset` or `vk::offset` using this
+type to unambiguously lay out these objects.
+
+## Motivation
+
+HLSL has a few contexts where we have types with a layout that doesn't match
+the usual rules that follow from C++ definitions and targets' data layouts. We
+can generally describe the appropriate type representations for these using
+explicit padding, but some care needs to be taken:
+
+- Arrays in CBuffers may have padding in between members, but crucially they do
+  not have padding after the last member. This needs special handling to
+  represent.
+
+- There are two HLSL constructs that may introduce arbitrary padding to a
+  struct. In a cbuffer, the `packoffset` attribute specifies the offset of a
+  member, and outside of cbuffers, the `vk::offset` vulkan attribute may do the
+  same.
+
+- Simply padding structures with `i8` as is typical with ABI-related padding
+  makes it difficult to recover which struct elements are padding vs which are
+  subobjects. This matters in some backends, and is specifically important for
+  SPIR-V where we need to map a logical indices into the struct into physical
+  offsets.
+
+## Proposed solution
+
+Introduce an explicit padding type and use this type for the padding in the
+various constructs that need it.
+
+The padding type will be a target type with a parameter for the size in bytes.
+For DirectX, this would look like `target("dx.Layout", 4)` for 4 bytes of
+padding. For SPIR-V, `target("spirv.Padding", 8)` for 8 bytes of padding. We
+may attempt to come up with a first-class type in LLVM for these purposes in
+the future.
+
+### Structs with annotations
```

Comment:
I would probably move this section after "### Arrays in a cbuffer" because this is way less common than ararys in cbuffer.